src/ssh-util/src/tunnel.rs

```diff
@@ -110,33 +127,66 @@ impl SshTunnelConfig {
                 scopeguard::defer! {
                     info!(%tunnel_id, "terminating ssh tunnel");
                 }
+                let mut interval = time::interval(CHECK_INTERVAL);
+                // Just in case checking takes a long time.
+                interval.set_missed_tick_behavior(time::MissedTickBehavior::Skip);
+                // The first tick happens immediately.
+                interval.tick().await;
                 loop {
-                    time::sleep(CHECK_INTERVAL).await;
+                    let closed = matches!(
+                        status_rx.try_recv(),
+                        Err(tokio::sync::mpsc::error::TryRecvError::Disconnected)
+                    );
+
+                    use mz_ore::channel::ReceiverExt;
+                    let interested_in_status = tokio::select! {
+                        _ = interval.tick() => {
+                            Vec::new()
+                        }
+                        Some(senders) = status_rx.recv_many(100), if !closed => {
+                            senders
+                        }
+                    };
+
                     if let Err(e) = session.check().await {
                         warn!(%tunnel_id, "ssh tunnel unhealthy: {}", e.display_with_causes());
                         let s = match connect(&config).await {
                             Ok(s) => s,
                             Err(e) => {
                                 warn!(%tunnel_id, "reconnection to ssh tunnel failed: {}", e.display_with_causes());
+
+                                for interested in interested_in_status {
+                                    let _ = interested
+                                        .send(SshTunnelStatus::Errored(e.to_string_with_causes()));
+                                }
+
                                 continue;
                             }
                         };
                         let lp = match port_forward(&s, &remote_host, remote_port).await {
                             Ok(lp) => lp,
                             Err(e) => {
                                 warn!(%tunnel_id, "reconnection to ssh tunnel failed: {}", e.display_with_causes());
+                                for interested in interested_in_status {
+                                    let _ = interested
+                                        .send(SshTunnelStatus::Errored(e.to_string_with_causes()));
+                                }
                                 continue;
                             }
                         };
                         session = s;
-                        local_port.store(lp, Ordering::SeqCst)
+                        local_port.store(lp, Ordering::SeqCst);
+                    }
+                    for interested in interested_in_status {
+                        let _ = interested.send(SshTunnelStatus::Running);
```

Comment:
Struggling to follow the channel sending code here. What does this buy us over `Arc<Mutex<SshTunnelStatus>>` that's set whenever an error occurs and cleared after a successful connection? Waiting out a tick interval? That seems perhaps undesirable? I feel like just having `check_status()` return the state of the SSH tunnel from the last check would be sufficient and simpler!

Reply:
@benesch it ensures we get an up-to-date status every time we ask; otherwise the kafka source could check the status, get the last running status of Running, and then thing ssh is fine, when it just failed. working around that any other way, like with an `Arc<Mutex>`, which i tried in the original version of this pr months ago was a fiasco

Reply:
Sorry I wasn't more clear initially, but I am uncomfortable with this level of complexity and think we need to find a simpler way!


> ; otherwise the kafka source could check the status, get the last running status of Running, and then thing ssh is fine, when it just failed

That's true here though too, no? The source could ask for a status, wait for a check, get "Ok", and then the SSH tunnel could immediately fail. I don't think waiting out a check interval buys us anything?

What goes wrong with the `Arc<Mutex<_>>`? It seems like it would be quite straightforward!

Reply:
the callee still needs a way to be told when a check finishes, which means either a broadcast channel, a watch channel, or a `mpsc<oneshot>` like we have here; I don't there there is a way to avoid a channel of some kind